bindings/netstandard/ElectionGuard/ElectionGuard.Decryption/Extensions/ComputeChaumPedersenProofs.cs

```diff
@@ -0,0 +1,228 @@
+        List<GuardianChallengeResponse> responses,
+        bool skipValidation = false
+    )
+    {
+        self.AddProofs(
+             tally,
+             responses.Select(x => x.Tally).ToList(),
+             skipValidation
+         );
+    }
+
+    /// <summary>
+    /// Add proofs to the tally
+    /// </summary>
+    public static void AddProofs(
+        this AccumulatedTally self,
+        CiphertextTally tally,
+        List<TallyChallengeResponse> responses,
+        bool skipValidation = false
+    )
+    {
+        foreach (var (contestId, contest) in self.Contests)
+        {
+            var contestResponses = responses.Select(
+                x => x.Contests[contestId]).ToList();
+            contest.AddProofs(
+                tally.Contests[contestId],
+                tally.Context,
+                contestResponses,
+                skipValidation
+            );
+        }
+    }
+
+    /// <summary>
+    /// Add proofs to the collection of ballots
+    /// </summary>
+    public static void AddProofs(
+       this List<AccumulatedBallot> self,
+       List<CiphertextBallot> ballots,
+       CiphertextElectionContext context,
+       List<GuardianChallengeResponse> responses,
+       bool skipValidation = false
+   )
+    {
+        var ballotResponses = new Dictionary<string, List<BallotChallengeResponse>>();
+        foreach (var response in responses)
+        {
+            foreach (var ballotChallengeResponse in response.Ballots)
+            {
+                if (!ballotResponses.ContainsKey(ballotChallengeResponse.ObjectId))
+                {
+                    ballotResponses.Add(ballotChallengeResponse.ObjectId, new List<BallotChallengeResponse>());
+                }
+
+                ballotResponses[ballotChallengeResponse.ObjectId].Add(ballotChallengeResponse);
+            }
+        }
+
+        self.AddProofs(ballots, context, ballotResponses, skipValidation);
+    }
+
+    /// <summary>
+    /// Add proofs to the collection of ballots
+    /// </summary>
+    public static void AddProofs(
+        this List<AccumulatedBallot> self,
+        List<CiphertextBallot> ballots,
+        CiphertextElectionContext context,
+        Dictionary<string, List<BallotChallengeResponse>> responses,
+        bool skipValidation = false
+    )
+    {
+        for (var i = 0; i < self.Count; i++)
+        {
+            var ballot = ballots[i];
+            var ballotResponses = responses[ballot.ObjectId];
+            self[i].AddProofs(
+                ballot, context, ballotResponses, skipValidation);
+        }
+    }
+
+    /// <summary>
+    /// Add proofs to the ballot
+    /// </summary>
+    public static void AddProofs(
+        this AccumulatedBallot self,
+        CiphertextBallot ballot,
+        CiphertextElectionContext context,
+        List<BallotChallengeResponse> responses,
+        bool skipValidation = false
+    )
+    {
+        foreach (var (contestId, contest) in self.Contests)
```

Comment:
Any reason that this one was a foreach and the one above a for?

Reply:
just because of the multi-collection item lookups done in the method abovce which works over 3 different collections. it felt more legible that way rather than having to use linq for both ballot and ballotResponse